dig_test.go

```diff
@@ -0,0 +1,360 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestEndToEndSuccess(t *testing.T) {
+	t.Parallel()
+
+	t.Run("pointer", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(&bytes.Buffer{}), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.NotNil(t, b, "invoke got nil buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("nil pointer", func(t *testing.T) {
+		// Dig shouldn't forbid this - it's perfectly reasonable to explicitly
+		// provide a typed nil, since that's often a convenient way to supply a
+		// default no-op implementation.
+		c := New()
+		require.NoError(t, c.Provide((*bytes.Buffer)(nil)), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.Nil(t, b, "expected to get nil buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("pointer constructor", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() *bytes.Buffer {
+			return &bytes.Buffer{}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.NotNil(t, b, "invoke got nil buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("struct", func(t *testing.T) {
+		t.Skip("Not yet supported.")
+		c := New()
+		buf := &bytes.Buffer{}
+		buf.WriteString("foo")
+		require.NoError(t, c.Provide(*buf), "provide failed")
+		require.NoError(t, c.Invoke(func(b bytes.Buffer) {
+			// ensure we're getting back the buffer we put in
+			require.Equal(t, "foo", buf.String(), "invoke got new buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("struct constructor", func(t *testing.T) {
+		t.Skip("Not yet supported.")
+		c := New()
+		buf := &bytes.Buffer{}
+		buf.WriteString("foo")
+		require.NoError(t, c.Provide(*buf), "provide failed")
+		require.NoError(t, c.Invoke(func(b bytes.Buffer) {
+			// ensure we're getting back the buffer we put in
+			require.Equal(t, "foo", buf.String(), "invoke got new buffer")
+		}), "invoke failed")
+	})
+
+	t.Run("slice", func(t *testing.T) {
+		c := New()
+		bufs := []*bytes.Buffer{{}, {}}
+		require.NoError(t, c.Provide(bufs), "provide failed")
+		require.NoError(t, c.Invoke(func(bs []*bytes.Buffer) {
+			require.Equal(t, 2, len(bs), "invoke got unexpected number of buffers")
+		}), "invoke failed")
+	})
+
+	t.Run("slice constructor", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() []*bytes.Buffer {
+			return []*bytes.Buffer{{}, {}}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(bs []*bytes.Buffer) {
+			require.Equal(t, 2, len(bs), "invoke got unexpected number of buffers")
+		}), "invoke failed")
+	})
+
+	t.Run("array", func(t *testing.T) {
+		c := New()
+		bufs := [1]*bytes.Buffer{{}}
+		require.NoError(t, c.Provide(bufs), "provide failed")
+		require.NoError(t, c.Invoke(func(bs [1]*bytes.Buffer) {
```

Comment:
IMO, we want to work with whatever is provided. It's up to the user to decide which they want.